includes/gateway/class-omise-payment-promptpay.php

```diff
@@ -0,0 +1,207 @@
+
+/**
+ * @since 4.1
+ */
+class Omise_Payment_Promptpay extends Omise_Payment_Offline {
+	public function __construct() {
+		parent::__construct();
+
+		$this->id                 = 'omise_promptpay';
+		$this->has_fields         = false;
+		$this->method_title       = __( 'Omise PromptPay', 'omise' );
+		$this->method_description = wp_kses(
+			__( 'Accept payments through <strong>PromptPay</strong> via Omise payment gateway.', 'omise' ),
+			array( 'strong' => array() )
+		);
+
+		$this->init_form_fields();
+		$this->init_settings();
+
+		$this->title                = $this->get_option( 'title' );
+		$this->description          = $this->get_option( 'description' );
+		$this->restricted_countries = array( 'TH' );
+		$this->source_type          = 'promptpay';
+
+		add_action( 'woocommerce_update_options_payment_gateways_' . $this->id, array( $this, 'process_admin_options' ) );
+		add_action( 'woocommerce_order_action_' . $this->id . '_sync_payment', array( $this, 'sync_payment' ) );
+		add_action( 'woocommerce_thankyou_' . $this->id, array( $this, 'display_qrcode' ) );
+		add_action( 'woocommerce_email_after_order_table', array( $this, 'email_qrcode' ) );
+	}
+
+	/**
+	 * @see WC_Settings_API::init_form_fields()
+	 * @see woocommerce/includes/abstracts/abstract-wc-settings-api.php
+	 */
+	public function init_form_fields() {
+		$this->form_fields = array(
+			'enabled' => array(
+				'title'   => __( 'Enable/Disable', 'omise' ),
+				'type'    => 'checkbox',
+				'label'   => __( 'Enable Omise PromptPay Payment', 'omise' ),
+				'default' => 'no'
+			),
+
+			'title' => array(
+				'title'       => __( 'Title', 'omise' ),
+				'type'        => 'text',
+				'description' => __( 'This controls the title the user sees during checkout.', 'omise' ),
+				'default'     => __( 'PromptPay', 'omise' ),
+			),
+
+			'description' => array(
+				'title'       => __( 'Description', 'omise' ),
+				'type'        => 'textarea',
+				'description' => __( 'This controls the description the user sees during checkout.', 'omise' )
+			),
+		);
+	}
+
+	/**
+	 * @param WC_Order $order
+	 *
+	 * @see   woocommerce/templates/emails/email-order-details.php
+	 * @see   woocommerce/templates/emails/plain/email-order-details.php
+	 */
+	public function email_qrcode( $order ) {
+		if ( $this->id == $order->get_payment_method() ) {
+			$this->display_qrcode( $order, 'email' );
+		}
+	}
+
+	/**
+	 * @param int|WC_Order $order
+	 * @param string       $context  pass 'email' value through this argument only for 'sending out an email' case.
+	 */
+	public function display_qrcode( $order, $context = 'view' ) {
+		if ( ! $this->load_order( $order ) ) {
+			return;
+		}
+
+		$charge = OmiseCharge::retrieve( $this->get_charge_id_from_order() );
+		$qrcode = $charge['source']['scannable_code']['image']['download_uri'];
+
+		$expires_datetime = new WC_DateTime( $charge['expires_at'], new DateTimeZone( 'UTC' ) );
+		$expires_datetime->set_utc_offset( wc_timezone_offset() );
```

Comment:
Note: there is a `timezone-offset` configuration in WordPress itself. This part is to make sure that the plugin will set the time according to the WordPress configuration instead of server time/timezone.